### DISCUSSION TITLE
asp core 3.1 set global default timezone?

### DISCUSSION THREAD

**Discussion by EvgenyPrikhodko** [dotnet/runtime#42153] [General] [answered]
How to specify the global timzone?

**Comment by ericstj** [comment]
cc @safern @tarekgh @krwq

**Comment by tarekgh** [comment, marked as answer]
It has to be done through the User settings or the machine admin. We don't have API to set the default time zone on the machine because it need a specific privilege to do it which is turned off by default.  

As user if you want to change the TZ from the settings you may follow the steps https://support.microsoft.com/en-us/help/4026213/windows-how-to-set-your-time-and-time-zone. 

If you want to do it by APIs, you may consider pinvoke to https://docs.microsoft.com/en-us/windows/win32/api/timezoneapi/nf-timezoneapi-settimezoneinformation but you have to work getting the needed privlage before doing so or the API will fail.

**Reply by tarekgh** [reply]
Also, I want to point, changing the time zone by API would cause a problem if the change is not desired by the machine user which can affect many other running applications. So if you do that you have to be very careful and make sure will not cause any problem.

**Comment by EvgenyPrikhodko** [comment]
There is a backend API (asp netCore 3.1 + EFCore) works on centos-> kestrel -> nginx. 
Backend run as service (deamon), OS timezone x, API time zone y.
How to run API timezone != TimezoneOS?

**Comment by tarekgh** [comment]
@EvgenyPrikhodko what is your scenario that need to run with different time zone than TimezoneOS? I am not familiar with what `asp netcore3.1 + EFCore` offering. CC @davidfowl if he knows more about that.

If you want to do that for testing purpose, one way to do it is using the Reflection to set the field https://source.dot.net/#System.Private.CoreLib/TimeZoneInfo.cs,65.

**Reply by EvgenyPrikhodko** [reply]
@tarekgh  `DateTime.Now`

**Reply by tarekgh** [reply]
Is this just for testing? or this is needed in your product? Also, I can provide you a code which can get the DateTime.Now in any other time zone.

**Reply by EvgenyPrikhodko** [reply]
Unfortunately, there is too much code, for example using `TimeZoneInfo` too big refactoring

**Reply by tarekgh** [reply]
Here is the code example, just in case you decide to use it. Maybe searching and replacing `DateTime.Now` can be simple. 

```C#
        static TimeZoneInfo s_tzi = TimeZoneInfo.FindSystemTimeZoneById("Morocco Standard Time");
        
        // ...

        DateTime dateTimeNow = TimeZoneInfo.ConvertTimeFromUtc(DateTime.UtcNow, s_tzi);
```

As I mention, the other idea would be use the reflection to set https://source.dot.net/#System.Private.CoreLib/TimeZoneInfo.cs,65

**Reply by EvgenyPrikhodko** [reply]
Thank you very much, the problem is that you have to rewrite too much code.

**Reply by EvgenyPrikhodko** [reply]
[Hoped to solve the problem somehow like this](https://github.com/OData/WebApi/issues/1238#issuecomment-365165199)

**Reply by EvgenyPrikhodko** [reply]
Is there no such feature in Net Core?

**Reply by tarekgh** [reply]
No. I am not even sure how you want this feature work? override the Local time zone for the whole process so it reflect to all code calling DateTime.Now? 

Even I looked at the code of `AspNet.OData` and I am seeing they having just a wrapper class. https://github.com/shubhlive/WEBAPIPackt/blob/master/src/Microsoft.AspNet.OData.Shared/TimeZoneInfoHelper.cs and then they have the extension method https://github.com/shubhlive/WEBAPIPackt/blob/e45702c1320083af8e16ce95bda953e6c961c014/src/Microsoft.AspNetCore.OData/Extensions/ODataRouteBuilderExtensions.cs#L405 to get the time zone. It is not really overriding the local time zone but just carrying some TimeZoneInfo object which they use for time conversion. DateTime.Now in that scenario will still be broken I guess.

**Reply by EvgenyPrikhodko** [reply]
Override the Local time zone for the whole process so it reflect to all code calling DateTime.Now?
Yes, It is possible?

**Reply by EvgenyPrikhodko** [reply]
use-case
For example, many different applications are running on the same operating system, but separately they should work in different timzones, without global rewriting of the code.

**Reply by tarekgh** [reply]
Officially we don't have such feature but you can achieve that by using the reflection to set https://source.dot.net/#System.Private.CoreLib/TimeZoneInfo.cs,65. Let me know if you want some code example how to do that. I want to point this will be not supported way as things can change in the future but it is very unlikely to change for the .NET version you are already using.

**Reply by EvgenyPrikhodko** [reply]
Thank you for your feedback, I understand, I will be glad if you can help in this matter.

**Reply by EvgenyPrikhodko** [reply]
> Officially we don't have such feature but you can achieve that by using the reflection to set https://source.dot.net/#System.Private.CoreLib/TimeZoneInfo.cs,65. Let me know if you want some code example how to do that. I want to point this will be not supported way as things can change in the future but it is very unlikely to change for the .NET version you are already using.

Please show an example of your idea

**Reply by tarekgh** [reply]
I'll try to work getting this for you. give me some time to do so. Thanks.

**Reply by tarekgh** [reply]
Here is some code demonstrate how to set the local time zone:

```C#
        public static bool SetLocalTimeZone(TimeZoneInfo local)
        {
            FieldInfo cachedDataField = typeof(TimeZoneInfo).GetField("s_cachedData", BindingFlags.NonPublic | BindingFlags.Static);
            if (cachedDataField == null)
            {
                return false;
            }

            object cachedData = cachedDataField.GetValue(null);
            if (cachedData == null)
            {
                return false;
            }

            FieldInfo localTimeZone = typeof(TimeZoneInfo).GetNestedType("CachedData", BindingFlags.NonPublic).GetField("m_localTimeZone", BindingFlags.NonPublic | BindingFlags.Instance) ??
                                      typeof(TimeZoneInfo).GetNestedType("CachedData", BindingFlags.NonPublic).GetField("_localTimeZone", BindingFlags.NonPublic | BindingFlags.Instance);
            if (localTimeZone == null)
            {
                return false;
            }

            localTimeZone.SetValue(cachedData, local);

            // The following code is for Windows, so if it fail that is ok as we may be running on Linux

            Type offsetAndRuleType = typeof(TimeZoneInfo).GetNestedType("OffsetAndRule", BindingFlags.NonPublic);
            if (offsetAndRuleType != null)
            {
                ConstructorInfo [] cInfo = offsetAndRuleType.GetConstructors(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance);
                if (cInfo != null && cInfo.Length == 1)
                {
                    TimeZoneInfo.AdjustmentRule rule = null;
                    TimeZoneInfo.AdjustmentRule[] rules = local.GetAdjustmentRules();
                    int year = DateTime.Now.Year;
                    if (rules != null)
                    {
                        foreach (var r in rules)
                        {
                            if (r.DateStart.Year == year)
                            {
                                rule = r;
                                break;
                            }
                        }
                    }

                    object o = cInfo[0].Invoke(new object[] { year, local.BaseUtcOffset, rule });

                    FieldInfo oneYearLocalFromUtc = typeof(TimeZoneInfo).GetNestedType("CachedData", BindingFlags.NonPublic).GetField("m_oneYearLocalFromUtc", BindingFlags.NonPublic | BindingFlags.Instance) ??
                                                    typeof(TimeZoneInfo).GetNestedType("CachedData", BindingFlags.NonPublic).GetField("_oneYearLocalFromUtc", BindingFlags.NonPublic | BindingFlags.Instance);
                    if (oneYearLocalFromUtc != null)
                    {
                        oneYearLocalFromUtc.SetValue(cachedData, o);
                    }
                }
            }

            return true;
        }
```

if you run a code like the following: 

```C#
            Console.WriteLine($"TZ: {TimeZoneInfo.Local.StandardName} ... {DateTime.Now}");

            SetLocalTimeZone(TimeZoneInfo.FindSystemTimeZoneById("Morocco Standard Time"));
            // On Linux: SetLocalTimeZone(TimeZoneInfo.FindSystemTimeZoneById("Africa/Casablanca"));

            Console.WriteLine($"TZ: {TimeZoneInfo.Local.StandardName} ... {DateTime.Now}");
```

will get output like:


```
TZ: Pacific Standard Time ... 9/14/2020 9:02:24 PM
TZ: Morocco Standard Time ... 9/15/2020 5:02:24 AM
```

## Warning ##

Although this code will work (I already tried it on Windows and Linux, both .NET Framework and .Net Core), but it is fragile because it can rest when any one calls `TimeZoneInfo.ClearCachedData`, this is unlikely but it can happen. Also the logic can reset again on Windows if the time cross the year i.e. crossing Dec 31st to Jan 1st.

**Reply by EvgenyPrikhodko** [reply]
Thank you very much for the sample code, we will test it.
1. Can this Method be called once in Main for the entire Project?
2. Logic can reset again on Windows, no problem for Linux?

**Reply by tarekgh** [reply]
> Can this Method be called once in Main for the entire Project?

Yes. 

> Logic can reset again on Windows, no problem for Linux?

Yes. When the time cross the year the rest will happen only on Windows.

But if anyone calls `TimeZoneInfo.ClearCachedData` the reset will happen on Windows and Linux. you may try to play with that to see the behavior.

**Reply by EvgenyPrikhodko** [reply]
> > Can this Method be called once in Main for the entire Project?
> 
> Yes.
> 
> > Logic can reset again on Windows, no problem for Linux?
> 
> Yes. When the time cross the year the rest will happen only on Windows.
> 
> But if anyone calls `TimeZoneInfo.ClearCachedData` the reset will happen on Windows and Linux. you may try to play with that to see the behavior.

Thank you so much for your answers.

**Reply by tarekgh** [reply]
I want to emphasize you need to test this well with your scenario before you apply it. As I mentioned this is fragile and can break. The provided code is just example and not intended to be used in real product but it is your decision if you want to use it and depend on it.

**Reply by fukasawah** [reply]
This answer was very helpful to me.
I develop on Windows and run on Linux, and in both cases I needed to make TimeZoneInfo.Local behave like UTC time.

``` cs
SetLocalTimeZone(TimeZoneInfo.CreateCustomTimeZone(
    "UTC-FOR-TEST", TimeZoneInfo.Utc.BaseUtcOffset, TimeZoneInfo.Utc.DisplayName, TimeZoneInfo.Utc.StandardName
));
```

However, if I call TimeZoneInfo.ConvertTime when datetime.Ticks==0, it calls ClearCachedData.

https://github.com/dotnet/runtime/blob/215b39abf947da7a40b0cb137eab4bceb24ad3e3/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs#L605-L613

It consumed a lot of time until I noticed this behavior...

**Comment by EvgenyPrikhodko** [comment]
please register my wish.
use-case
For example, I have an application that was written without timzone, I don't know what timezone hosting uses (azure, amazone, etc.).
Make `dotnet myapp.dll --timezone "Morocco Standard Time"`

**Reply by EvgenyPrikhodko** [reply]
@tarekgh Please consider my offer

**Reply by tarekgh** [reply]
This need to be turned into a proposal and logged a new issue as a request. The proposal has to be detailed for the behavior and if there is any needed new API. Here is some example https://github.com/dotnet/runtime/issues/42109

**Reply by EvgenyPrikhodko** [reply]
[Create](https://github.com/dotnet/runtime/issues/42241)

**Comment by PathogenDavid** [comment]
I'm a bit late, but if you only need Linux support for this, .NET respects the `TZ` environment variable as shown below:

![image](https://user-images.githubusercontent.com/278957/98411686-2244e880-203c-11eb-8766-b1fb9bac9b3e.png)

The value of TZ can basically be any file under `/usr/share/zoneinfo`.

**Reply by SteveDesmond-ca** [reply]
This just saved me a ton of time zone logic code, thanks @PathogenDavid!

**Comment by tarekgh** [comment]
@PathogenDavid what is your scenario you want to use this for?

**Reply by PathogenDavid** [reply]
[Replied on #42241](https://github.com/dotnet/runtime/issues/42241#issuecomment-723500292)
